### DISCUSSION TITLE
How is battery shutdown handled?

### DISCUSSION THREAD

**Discussion by Novakasa** [pybricks/support#1124] [Q&A] [answered]
I am currently wondering, due to https://github.com/Novakasa/brickrail/issues/140, how battery shutdown is handled with pybricks.
According to the user log the program was just stopped normally (pybricksdev `run` coroutine returns without error), and I am wondering whether that could've been due to battery shutdown. Is a voltage of 7V enough to trigger a battery shutdown on a cityhub? Is there an easy way to detect impending battery shutdown so I can notify the user?

**Comment by dlech** [comment, marked as answer]
> Is a voltage of 7V enough to trigger a battery shutdown on a cityhub?

No, it doesn't shut down until 4.8V. https://github.com/pybricks/pybricks-micropython/blob/master/lib/pbio/sys/battery.c



> Is there an easy way to detect impending battery shutdown so I can notify the user?

If you have a Bluetooth connection, the status events include a low battery voltage warning flag.

**Reply by Novakasa** [reply]
Thanks. Yeah after thinking about it for 1 second, rechargeable batteries only produce 7.2 V :sweat_smile: 

> If you have a Bluetooth connection, the status events include a low battery voltage warning flag

That's very useful to know, thanks.

**Comment by BertLindeman** [comment]
The hubs (I think all of them) have a battery.voltage() function.

An example:
```python
from pybricks.hubs import CityHub
from pybricks.tools import wait

hub = CityHub()

# from doc: "Using the battery"
#   battery.voltage()→ int: mV

for i in range(10):
    print(hub.battery.voltage())
    wait(1000)
```

Sample output:
```

7531
7532
7532
7532
7532
7532
7531
7531
7531
7530
```

I am using something similar to check the battery of my EV3 to  warn using the leds to flash yellow or orange or red depending on the voltage.
